downstairs/src/region.rs

```diff
@@ -1754,14 +1792,15 @@ impl Region {
      *   D. Only then, open extent.
      */
     pub async fn repair_extent(
-        &mut self,
+        &self,
         eid: usize,
         repair_addr: SocketAddr,
     ) -> Result<(), CrucibleError> {
         // Make sure the extent:
         // is currently closed, matches our eid, is not read-only
-        assert!(self.extents[eid].inner.is_none());
-        assert_eq!(self.extents[eid].number, eid as u32);
+        let mg = self.extents[eid].lock().await;
```

Comment:
What happens if we are repairing extents and another downstairs takes over?

I'm not sure if have any tests for this, but.  hmm.  What would we need to to to
create a test of this sort of situation to be sure we remain crash consistent here?
I just wonder about the lock and preventing two upstairs connections from making
a mess of things.

Reply:
I wrote one in 4955d17 - it seems like the old upstairs that's in the middle of repairing gets deactivated, the new upstairs tries to activate, then repairs the downstairs during activation due to the flush numbers mismatch.

Reply:
That's the behavior I would want, but what I really want is to run that test XX times and make sure
we don't have any race conditions or other cases that we are not seeing here.

I'm just not sure we have protected ourselves for this case.  Granted it's not likely to come
up, but I just don't know what we do during repair if another upstairs takes over.  Maybe it
is all okay and the takeover upstairs does the right thing by waiting for the other tasks for
the other upstairs (that is now no longer active) to be finished with it's work.

I believe there is some logic in the work queue for the downstairs that check for this
situation and self destructs if it finds it.?

Reply:
> Maybe it is all okay and the takeover upstairs does the right thing by waiting for the other tasks for the other upstairs (that is now no longer active) to be finished with it's work.

I don't think this is the case: the old Upstairs receives the YouAreNoLongerActive and terminates itself in the middle of the repair. I tested this with 

```diff
diff --git a/upstairs/src/dummy_downstairs_tests.rs b/upstairs/src/dummy_downstairs_tests.rs
index 8962ad3..b58d23f 100644
--- a/upstairs/src/dummy_downstairs_tests.rs
+++ b/upstairs/src/dummy_downstairs_tests.rs
@@ -1069,6 +1069,20 @@ pub(crate) mod protocol_test {
                 }
             };
 
+            if eid == 5 {
+                // Send a YouAreNoLongerActive to nuke the Upstairs
+                harness.ds2.fw
+                    .lock()
+                    .await
+                    .send(Message::YouAreNoLongerActive {
+                        new_upstairs_id: Uuid::new_v4(),
+                        new_session_id: Uuid::new_v4(),
+                        new_gen: 9876,
+                    })
+                    .await
+                    .unwrap();
+            }
+
             // Extent limit is Some(eid), where eid is the current loop
             // iteration. It marks the extent at and below are clear to receive
             // IO. Issue some single extent reads and writes to make sure that
@@ -1081,6 +1095,7 @@ pub(crate) mod protocol_test {
 
             let mut responses = vec![Vec::new(); 3];
 
+            /*
             for io_eid in 0usize..10 {
                 // read
 
@@ -1280,6 +1295,7 @@ pub(crate) mod protocol_test {
                     _ => panic!("saw {:?}", m3),
                 }
             }
+            */
 
             // The repair task then waits for the close responses.
 

```

in the middle of `dummy_downstairs_tests::protocol_test::test_successful_live_repair` and progress stopped:

```
Jun 19 15:22:11.323 INFO received ExtentLiveClose { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2221, dependencies: [], extent_id: 5 }, downstairs: 1
Jun 19 15:22:11.323 INFO received ExtentLiveFlushClose { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2221, dependencies: [], extent_id: 5, flush_number: 7, gen_number: 1 }, downstairs: 2
Jun 19 15:22:11.323 INFO received ExtentLiveFlushClose { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2221, dependencies: [], extent_id: 5, flush_number: 7, gen_number: 1 }, downstairs: 3
Jun 19 15:22:11.366 INFO received ExtentLiveReopen { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2224, dependencies: [2221, 2222, 2223], extent_id: 5 }, downstairs: 3
Jun 19 15:22:11.366 INFO received ExtentLiveReopen { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2224, dependencies: [2221, 2222, 2223], extent_id: 5 }, downstairs: 2
Jun 19 15:22:11.366 INFO received ExtentLiveReopen { upstairs_id: c4752b14-c3e0-49a8-89b5-317ed16cd02d, session_id: 29d863af-b9d3-4b97-9a36-bea3c630d997, job_id: 2224, dependencies: [2221, 2222, 2223], extent_id: 5 }, downstairs: 1
Jun 19 15:22:11.367 ERRO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d (29d863af-b9d3-4b97-9a36-bea3c630d997) cmd_loop saw YouAreNoLongerActive 82a545bb-2df4-4ba4-a62a-c16fb6947f16 adcbfbbb-72f3-426b-af21-c8bb8ae2b729 9876, upstairs: 1
Jun 19 15:22:11.367 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d (29d863af-b9d3-4b97-9a36-bea3c630d997) LiveRepair Active Active ds_transition to Disabled, upstairs: 1
Jun 19 15:22:11.367 INFO [1] Transition from Active to Disabled, upstairs: 1
Jun 19 15:22:11.367 INFO c4752b14-c3e0-49a8-89b5-317ed16cd02d set inactive, session 29d863af-b9d3-4b97-9a36-bea3c630d997, upstairs: 1
Jun 19 15:22:11.367 ERRO 127.0.0.1:40203: proc: Generation number is too low: saw YouAreNoLongerActive with larger gen 9876 than ours 1, looper: 1, upstairs: 1
Jun 19 15:22:11.367 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d Gone missing, transition from Disabled to Disconnected, upstairs: 1
Jun 19 15:22:11.367 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d connection to 127.0.0.1:40203 closed, looper: 1, upstairs: 1
Jun 19 15:22:11.367 WARN [1] pm_task rx.recv() is None, upstairs: 1
Jun 19 15:22:11.367 INFO [1] 127.0.0.1:40203 task reports connection:false, upstairs: 1
Jun 19 15:22:11.367 INFO c4752b14-c3e0-49a8-89b5-317ed16cd02d LiveRepair Disconnected Active, upstairs: 1
Jun 19 15:22:11.367 INFO [1] 127.0.0.1:40203 task reports offline, upstairs: 1
Jun 19 15:22:12.368 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d looper connected, looper: 1, upstairs: 1
Jun 19 15:22:12.368 INFO [1] Proc runs for 127.0.0.1:40203 in state Disconnected, upstairs: 1
Jun 19 15:23:02.369 ERRO 127.0.0.1:40203: proc: timed out during negotiation, looper: 1, upstairs: 1
Jun 19 15:23:02.369 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d Gone missing, transition from Disconnected to Disconnected, upstairs: 1
Jun 19 15:23:02.369 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d connection to 127.0.0.1:40203 closed, looper: 1, upstairs: 1
Jun 19 15:23:02.369 INFO [1] 127.0.0.1:40203 task reports connection:false, upstairs: 1
Jun 19 15:23:02.369 INFO c4752b14-c3e0-49a8-89b5-317ed16cd02d LiveRepair Disconnected Active, upstairs: 1
Jun 19 15:23:02.369 INFO [1] 127.0.0.1:40203 task reports offline, upstairs: 1
Jun 19 15:23:03.370 INFO [1] c4752b14-c3e0-49a8-89b5-317ed16cd02d looper connected, looper: 1, upstairs: 1
Jun 19 15:23:03.370 INFO [1] Proc runs for 127.0.0.1:40203 in state Disconnected, upstairs: 1
test dummy_downstairs_tests::protocol_test::test_successful_live_repair has been running for over 60 seconds
```

> I believe there is some logic in the work queue for the downstairs that check for this situation and self destructs if it finds it.?

You mean this?

https://github.com/oxidecomputer/crucible/blob/18d2fa49d716f62742d42fa9140cccbd0d5b894d/downstairs/src/lib.rs#L1988-L2001